File: go/回溯/47. 全排列 II.go
package main

import (
	"fmt"
	"sort"
)

var res [][]int

func permuteUnique(nums []int) [][]int {
	var path []int
	sort.Ints(nums)
	dfs(path, nums)
	return res
}

func dfs(path []int, choice []int) {
	if len(path) == len(choice) {
		res = append(res, path)
		return
	}
	for i := 0; i < len(choice); i++ {
		if i > 0 && choice[i] == choice[i-1] {
			continue
		}
		path = append(path, choice[i])
		//tmp := append(choice[:i], choice[i+1:]...)
		dfs(path, choice)
		//path = path[:len(choice)-1]
	}
}

func main() {
	nums := []int{1, 2, 3}
	fmt.Println(permuteUnique(nums))
}

File: python/general/47. 全排列 II.py
from typing import List

# 回朔法
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def dfs(path,nums,length):
            if len(path) == length:
                res.append(path[:])
                return
            for i in range(len(nums)):
                if i > 0 and nums[i] == nums[i-1]:continue
                path.append(nums[i])
                dfs(path,nums[:i]+nums[i+1:],length)
                path.pop()
        res = []
        length = len(nums)
        nums.sort()
        dfs([],nums,length)
        return res

# 不断获取下一个全排列

class Solution2:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        def nextPermutation(nums: List[int]) -> bool:
            i = len(nums) - 1
            while i > 0 and nums[i - 1] >= nums[i]:
                i -= 1
            if i <= 0:
                nums.sort()
                res.append(nums[:])
                return False

            j = len(nums) - 1
            while nums[i - 1] >= nums[j]:
                j -= 1
            nums[i - 1], nums[j] = nums[j], nums[i - 1]

            nums[i:] = nums[len(nums) - 1:i - 1:-1]

            res.append(nums[:])
            return True
        res = []
        nums.sort()
        while True:
            isTerminate = nextPermutation(nums)
            if not isTerminate:
                return res


s = Solution()
nums = [1,1,2]
print(s.permuteUnique(nums))


